telegram/ext/_updater.py

```diff
@@ -592,94 +596,44 @@ def bootstrap_onerr_cb(exc):
-        if self.dispatcher:
-            self.logger.debug('Requesting Dispatcher to stop...')
-            self.dispatcher.stop()
-
-    @no_type_check
-    def _join_threads(self) -> None:
-        for thr in self.__threads:
-            self.logger.debug('Waiting for %s thread to end', thr.name)
-            thr.join()
-            self.logger.debug('%s thread has ended', thr.name)
-        self.__threads = []
-
-    @no_type_check
-    def _signal_handler(self, signum, frame) -> None:
-        self.is_idle = False
-        if self.running:
-            self.logger.info(
-                'Received signal %s (%s), stopping...',
-                signum,
-                # signal.Signals is undocumented for some reason see
-                # https://github.com/python/typeshed/pull/555#issuecomment-247874222
-                # https://bugs.python.org/issue28206
-                signal.Signals(signum),  # pylint: disable=no-member
-            )
-            self.stop()
-            if self.user_signal_handler:
-                self.user_signal_handler(signum, frame)
-        else:
-            self.logger.warning('Exiting immediately!')
-            # pylint: disable=import-outside-toplevel, protected-access
-            import os
+        Raises:
+            :exc:`RuntimeError`: If the updater is not running.
+        """
+        async with self.__lock:
+            if not self.running:
+                raise RuntimeError('This Updater is not running!')
 
-            os._exit(1)
+            self._logger.debug('Stopping Updater')
 
-    def idle(
-        self, stop_signals: Union[List, Tuple] = (signal.SIGINT, signal.SIGTERM, signal.SIGABRT)
-    ) -> None:
-        """Blocks until one of the signals are received and stops the updater.
+            self._running = False
 
-        Args:
-            stop_signals (:obj:`list` | :obj:`tuple`): List containing signals from the signal
-                module that should be subscribed to. :meth:`Updater.stop()` will be called on
-                receiving one of those signals. Defaults to (``SIGINT``, ``SIGTERM``, ``SIGABRT``).
+            await self._stop_httpd()
+            await self._stop_polling()
 
-        """
-        for sig in stop_signals:
-            signal.signal(sig, self._signal_handler)
+            self._logger.debug('Updater.stop() is complete')
 
-        self.is_idle = True
+    async def _stop_httpd(self) -> None:
+        if self._httpd:
+            self._logger.debug('Waiting for current webhook connection to be closed.')
+            await self._httpd.shutdown()
+            self._httpd = None
 
-        while self.is_idle:
-            sleep(1)
+    async def _stop_polling(self) -> None:
+        if self.__polling_task:
+            self._logger.debug('Waiting background polling task to join.')
+            self.__polling_task.cancel()
+            try:
+                await self.__polling_task
+            except asyncio.CancelledError:
+                pass
```

Comment:
Will need to double check if this is one of those cases where asyncio complains about non-retrieved errors …

Reply:
okay, we really don't need the try-except because catching the `CancelledError` is done the network retry already. however, we do need to await: printing `__pooling.task.done()` directly after cancelling gives `False`

Reply:
edit: I was wrong. see the comment id added :D